src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_LENGTH_ARRAY = 10;
+    public static final int STARTING_ARRAY_INDEX = 0;
+    public static final double NUMBER_OF_GROWING_ARRAY_LENGTH = 1.5;
+    public static final int AMOUNT_OF_ADDED_VALUES_TO_ARRAY = 1;
+    public static final int AMOUNT_OF_REMOVED_VALUES_FROM_ARRAY = 1;
+    private T[] values;
+
+    private int size = 0;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_LENGTH_ARRAY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == values.length) {
+            grow();
+        }
+        addInList(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else {
+            checkIndex(index);
+            addInsideArray(value, index);
+        }
+
+    }
 
+    private void addInsideArray(T value, int index) {
+        if (size + AMOUNT_OF_ADDED_VALUES_TO_ARRAY == values.length) {
+            grow();
+        }
+        T[] tempValues = (T[]) new Object[values.length];
+        System.arraycopy(values, STARTING_ARRAY_INDEX, tempValues, STARTING_ARRAY_INDEX, index);
+        tempValues[index] = value;
+        System.arraycopy(values, index, tempValues, index + AMOUNT_OF_ADDED_VALUES_TO_ARRAY,
+                values.length - index - AMOUNT_OF_ADDED_VALUES_TO_ARRAY);
+        values = tempValues;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+
+    }
 
+    private void addInList(T value) {
+        values[size] = value;
+        size++;
+    }
+
+    private void grow() {
+        T[] grownArray = growFullArray();
+        System.arraycopy(values, STARTING_ARRAY_INDEX, grownArray, STARTING_ARRAY_INDEX, size);
+        values = grownArray;
+    }
+
+    private T[] growFullArray() {
+        return (T[]) new Object[(int) (values.length * NUMBER_OF_GROWING_ARRAY_LENGTH)];
     }
```

Comment:
You can omit this method, because you use it only once and it doesn't contains a lot of logic.
```suggestion
    private void grow() {
        T[] grownArray =  (T[]) new Object[(int) (values.length * NUMBER_OF_GROWING_ARRAY_LENGTH)];
        System.arraycopy(values, STARTING_ARRAY_INDEX, grownArray, STARTING_ARRAY_INDEX, size);
        values = grownArray;
    }
```